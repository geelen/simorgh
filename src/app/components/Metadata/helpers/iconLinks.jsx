import React from 'react';

export const getIconAssetUrl = (service, size) =>
  `https://news.files.bbci.co.uk/include/articles/public/${service}/images/icons/icon-${size}.png`;

const createIconLinks = (service, iconSizes, iconType) => {
  return iconSizes.map(size => {
    const iconAssetUrl = getIconAssetUrl(service, size);
    const key = `${service}-${size}`;

    if (iconType === 'icon') {
      return (
        <link
          key={key}
          rel="icon"
          type="image/png"
          href={iconAssetUrl}
          sizes={size}
        />
      );
    }

    return (
      <link key={key} rel="apple-touch-icon" sizes={size} href={iconAssetUrl} />
    );
  });
};

export const getIconLinks = (service, iconSizes) => {
  if (!iconSizes) {
    return null;
  }
  const iconTypes = Object.keys(iconSizes);
  return iconTypes.map(iconType => {
    return createIconLinks(service, iconSizes[iconType], iconType);
  });
};
