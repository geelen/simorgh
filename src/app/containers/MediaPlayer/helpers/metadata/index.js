import path from 'ramda/src/path';
import pathOr from 'ramda/src/pathOr';

const DEFAULT_IMAGE_RES = '1024x576';

export const getThumbnailUri = aresMetadataBlock => {
  const imageUrl = path(['model', 'imageUrl'], aresMetadataBlock);
  return `https://${imageUrl.replace('$recipe', DEFAULT_IMAGE_RES)}`;
};
export const getUploadDate = aresMetadataBlock => {
  const uploadDate = pathOr(
    null,
    ['model', 'versions', [0], 'availableFrom'],
    aresMetadataBlock,
  );
  return new Date(uploadDate).toISOString();
};

export const getType = aresMetadataBlock => {
  const format = path(['model', 'format'], aresMetadataBlock);
  return format === 'audio' ? 'AudioObject' : 'VideoObject';
};

export const getMetadata = aresMetadataBlock => {
  return {
    '@context': 'http://schema.org',
    '@type': getType(aresMetadataBlock),
    name: path(['model', 'title'], aresMetadataBlock),
    description: path(['model', 'synopses', 'short'], aresMetadataBlock),
    duration: pathOr(
      null,
      ['model', 'versions', [0], 'durationISO8601'],
      aresMetadataBlock,
    ),
    thumbnailUrl: getThumbnailUri(aresMetadataBlock),
    uploadDate: getUploadDate(aresMetadataBlock),
    embedURL: pathOr(null, ['embedSource'], aresMetadataBlock),
  };
};

export const getMetadataBlock = aresMediaBlocks => {
  const aresMetadataBlock = aresMediaBlocks.filter(
    block => block.type === 'aresMediaMetadata',
  );
  return aresMetadataBlock[0];
};

export const mediaPlayerMetadata = (aresMediaBlock, embedSource) =>
  getMetadata({ ...getMetadataBlock(aresMediaBlock), embedSource });
