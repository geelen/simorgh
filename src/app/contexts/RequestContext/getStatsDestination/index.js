/* Returns stats destnation for ATI based on origin, service and env
   see table on this issue https://github.com/bbc/simorgh/issues/2995
*/
const getStatsDestination = ({ isUK = true, env = 'test', service }) => {
  let destination = '';
  switch (service) {
    case 'news':
      // checks if "news" service is in the UK or not and set apprioprate destination
      destination = isUK !== false ? 'NEWS_PS' : 'NEWS_GNL';
      break;
    case 'cymrufyw':
    case 'naidheachdan':
      // checks if "cymrufyw" or "naidheachdan" service is in the UK or not and set apprioprate destination
      destination = isUK !== false ? 'NEWS_LANGUAGES_PS' : 'NEWS_LANGUAGES_GNL';
      break;
    case 'japanese':
      destination = 'NEWS_LANGUAGES_GNL';
      break;
    case 'scotland':
      destination = 'PLACEHOLDER';
      break;
    case 'archive':
      destination = 'BBC_ARCHIVE_PS';
      break;
    default:
      destination = 'WS_NEWS_LANGUAGES';
  }
  return env === 'live' ? destination : `${destination}_TEST`;
};

export default getStatsDestination;
