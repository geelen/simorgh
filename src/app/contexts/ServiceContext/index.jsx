import React from 'react';
import { node, string } from 'prop-types';
import services from '#lib/config/services/loadableConfig';
import { getVariant } from '../../lib/utilities/variantHandler';
import variantPropType from '../../models/propTypes/variants';

export const ServiceContext = React.createContext({});

export const ServiceContextProvider = ({ children, service, variant }) => {
  const LoadableContextProvider = services[service];

  if (!LoadableContextProvider) {
    return null;
  }

  return (
    <LoadableContextProvider
      Context={ServiceContext}
      dataKey={getVariant({ service, variant })}
    >
      {children}
    </LoadableContextProvider>
  );
};

ServiceContextProvider.propTypes = {
  children: node.isRequired,
  service: string,
  variant: variantPropType,
};

ServiceContextProvider.defaultProps = {
  service: 'default',
  variant: 'default',
};
