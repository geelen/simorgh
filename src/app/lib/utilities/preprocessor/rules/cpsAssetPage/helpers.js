export const optimoFragment = ({ fragment, attributes }) => ({
  type: 'fragment',
  model: {
    text: fragment,
    attributes,
  },
});

export const optimoParagraph = ({ fragments, text }) => ({
  type: 'paragraph',
  model: {
    text,
    blocks: fragments.map(optimoFragment),
  },
});

export const optimoText = paragraphs => ({
  type: 'text',
  model: {
    blocks: paragraphs.map(optimoParagraph),
  },
});
