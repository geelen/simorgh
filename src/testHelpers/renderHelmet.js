/*
 * Similar to this problem https://github.com/testing-library/react-testing-library/issues/402
 * This helper was created to solve the problem of rendering helmet/head contents in snapshots.
 * Pass in a component that uses helmet somewhere in the component tree.
 * The full html tree is rendered and returned which you can then use to snapshot helmet/head contents.
 * Example of use in a test file:

 it('should render correctly', async () => {
    const html = await renderHelmet(<SomeComponent {...someProps} />);
    expect(html).toMatchSnapshot();
  });

 */

import { render, waitForDomChange } from '@testing-library/react';

export default async component => {
  render(component);

  await waitForDomChange();

  const htmlElement = document.querySelector('html');

  const helmetElements = document.querySelectorAll(
    '[data-react-helmet="true"]',
  );

  const removeHelmetAttributes = el => el.removeAttribute('data-react-helmet'); // remove react-helmet attribute noise from elements

  removeHelmetAttributes(htmlElement); // remove react-helmet attribute noise from elements

  Array.from(helmetElements).forEach(removeHelmetAttributes);

  return document.querySelector('html');
};
