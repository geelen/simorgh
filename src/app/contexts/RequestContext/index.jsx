import React from 'react';
import { bool, node, oneOf, string, number } from 'prop-types';
import getStatsDestination from './getStatsDestination';
import getStatsPageIdentifier from './getStatsPageIdentifier';
import getOriginContext from './getOriginContext';
import getEnv from './getEnv';
import getMetaUrls from './getMetaUrls';
import variantPropType from '../../models/propTypes/variants';

export const RequestContext = React.createContext({});

export const RequestContextProvider = ({
  children,
  bbcOrigin,
  id,
  isAmp,
  pageType,
  service,
  statusCode,
  previousPath,
  pathname,
  variant,
}) => {
  const { isUK, origin } = getOriginContext(bbcOrigin);
  const env = getEnv(origin);
  const platform = isAmp ? 'amp' : 'canonical';
  const statsDestination = getStatsDestination({
    isUK,
    env,
    service,
  });
  const statsPageIdentifier = getStatsPageIdentifier({
    pageType,
    service,
    id,
  });

  const value = {
    env,
    id,
    isUK,
    origin,
    pageType,
    isAmp,
    platform,
    statsDestination,
    statsPageIdentifier,
    statusCode,
    previousPath,
    variant,
    ...getMetaUrls(origin, pathname),
  };

  return (
    <RequestContext.Provider value={value}>{children}</RequestContext.Provider>
  );
};

RequestContextProvider.propTypes = {
  bbcOrigin: string,
  children: node.isRequired,
  id: string,
  isAmp: bool.isRequired,
  pageType: oneOf([
    'article',
    'frontPage',
    'media',
    'error',
    'MAP',
    'FIX',
    'STY',
    'PGL',
  ]).isRequired,
  service: string.isRequired,
  statusCode: number,
  pathname: string.isRequired,
  previousPath: string,
  variant: variantPropType,
};

RequestContextProvider.defaultProps = {
  bbcOrigin: null,
  id: null,
  statusCode: null,
  previousPath: null,
  variant: null,
};
