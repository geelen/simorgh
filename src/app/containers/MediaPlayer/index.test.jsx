import React from 'react';
import { string, shape, node } from 'prop-types';
import { shouldMatchSnapshot } from '../../../testHelpers';
import MediaPlayerContainer from '.';
import { RequestContextProvider } from '../../contexts/RequestContext';
import { ToggleContext } from '../../contexts/ToggleContext';
import { validVideoFixture, missingVpidFixture } from './helpers/fixtures';

const defaultToggles = {
  test: {
    mediaPlayer: {
      enabled: true,
    },
  },
  live: {
    mediaPlayer: {
      enabled: false,
    },
  },
};

const mockToggleDispatch = jest.fn();
const ContextWrapper = ({ platform, children, toggleState }) => (
  <RequestContextProvider
    isAmp={platform === 'amp'}
    service="news"
    platform={platform}
    id="foo"
    pageType="article"
  >
    <ToggleContext.Provider
      value={{ toggleState, toggleDispatch: mockToggleDispatch }}
    >
      {children}
    </ToggleContext.Provider>
  </RequestContextProvider>
);

ContextWrapper.propTypes = {
  children: node.isRequired,
  platform: string.isRequired,
  toggleState: shape({}),
};

ContextWrapper.defaultProps = {
  toggleState: defaultToggles,
};

describe('MediaPlayer', () => {
  describe('is called correctly', () => {
    shouldMatchSnapshot(
      'Calls the canonical placeholder when platform is canonical',
      <ContextWrapper platform="canonical">
        <MediaPlayerContainer blocks={validVideoFixture} />
      </ContextWrapper>,
    );

    shouldMatchSnapshot(
      'Calls the canonical player when platform is canonical and placeholder is false',
      <ContextWrapper platform="canonical">
        <MediaPlayerContainer blocks={validVideoFixture} placeholder={false} />
      </ContextWrapper>,
    );

    shouldMatchSnapshot(
      'Calls the AMP player when platform is AMP',
      <ContextWrapper platform="amp">
        <MediaPlayerContainer blocks={validVideoFixture} />
      </ContextWrapper>,
    );
  });

  describe('Fails and returns early when', () => {
    shouldMatchSnapshot(
      'there is no versionId',
      <ContextWrapper platform="canonical">
        <MediaPlayerContainer blocks={missingVpidFixture} />
      </ContextWrapper>,
    );

    const toggleState = {
      test: {
        mediaPlayer: {
          enabled: false,
        },
      },
    };

    shouldMatchSnapshot(
      'component is toggled off',
      <ContextWrapper platform="canonical" toggleState={toggleState}>
        <MediaPlayerContainer blocks={validVideoFixture} />
      </ContextWrapper>,
    );
  });
});
