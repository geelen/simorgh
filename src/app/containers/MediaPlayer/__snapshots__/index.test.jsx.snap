// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaPlayer Fails and returns early when component is toggled off 1`] = `null`;

exports[`MediaPlayer Fails and returns early when there is no versionId 1`] = `null`;

exports[`MediaPlayer is called correctly Calls the AMP player when platform is AMP 1`] = `
.c1 {
  padding-top: 56.25%;
  position: relative;
  overflow: hidden;
}

.c0 {
  grid-column: 1 / span 6;
}

@media (max-width:25rem) {
  .c0 {
    padding: 0 0.5rem;
  }
}

@media (min-width:25rem) and (max-width:62.9375rem) {
  .c0 {
    padding: 0 1rem;
  }
}

@media (min-width:37.5rem) {
  .c0 {
    grid-column: 1 / span 5;
    max-width: 83.33%;
  }
}

@media (min-width:63rem) and (max-width:80rem) {
  .c0 {
    grid-column: 3 / span 5;
    max-width: 37.75rem;
  }
}

@media (min-width:80rem) {
  .c0 {
    grid-column: 6 / span 10;
    max-width: 38.5rem;
  }
}

@supports (display:grid) {
  .c0 {
    max-width: initial;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <amp-iframe
      allowfullscreen={true}
      frameborder="0"
      layout="fill"
      sandbox="allow-scripts allow-same-origin"
      src="https://www.test.bbc.com/ws/av-embeds/foo/p01k6msp/amp"
    >
      <amp-img
        layout="fill"
        placeholder={true}
        src="https://ichef.test.bbci.co.uk/images/ic/512xn/p01k6mtv.jpg"
      />
    </amp-iframe>
  </div>
</div>
`;

exports[`MediaPlayer is called correctly Calls the canonical placeholder when platform is canonical 1`] = `
.c3 {
  display: block;
  width: 100%;
  visibility: visible;
}

.c2 {
  cursor: pointer;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.c1 {
  padding-top: 56.25%;
  position: relative;
  overflow: hidden;
}

.c0 {
  grid-column: 1 / span 6;
}

@media (max-width:25rem) {
  .c0 {
    padding: 0 0.5rem;
  }
}

@media (min-width:25rem) and (max-width:62.9375rem) {
  .c0 {
    padding: 0 1rem;
  }
}

@media (min-width:37.5rem) {
  .c0 {
    grid-column: 1 / span 5;
    max-width: 83.33%;
  }
}

@media (min-width:63rem) and (max-width:80rem) {
  .c0 {
    grid-column: 3 / span 5;
    max-width: 37.75rem;
  }
}

@media (min-width:80rem) {
  .c0 {
    grid-column: 6 / span 10;
    max-width: 38.5rem;
  }
}

@supports (display:grid) {
  .c0 {
    max-width: initial;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
      onClick={[Function]}
    >
      <img
        alt="Image Alt"
        className="c3"
        height={null}
        sizes={null}
        src="https://ichef.test.bbci.co.uk/images/ic/512xn/p01k6mtv.jpg"
        srcSet={null}
        width={null}
      />
    </div>
  </div>
</div>
`;

exports[`MediaPlayer is called correctly Calls the canonical player when platform is canonical and placeholder is false 1`] = `
.c3 {
  display: block;
  width: 100%;
  visibility: visible;
}

.c2 {
  cursor: pointer;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.c1 {
  padding-top: 56.25%;
  position: relative;
  overflow: hidden;
}

.c0 {
  grid-column: 1 / span 6;
}

@media (max-width:25rem) {
  .c0 {
    padding: 0 0.5rem;
  }
}

@media (min-width:25rem) and (max-width:62.9375rem) {
  .c0 {
    padding: 0 1rem;
  }
}

@media (min-width:37.5rem) {
  .c0 {
    grid-column: 1 / span 5;
    max-width: 83.33%;
  }
}

@media (min-width:63rem) and (max-width:80rem) {
  .c0 {
    grid-column: 3 / span 5;
    max-width: 37.75rem;
  }
}

@media (min-width:80rem) {
  .c0 {
    grid-column: 6 / span 10;
    max-width: 38.5rem;
  }
}

@supports (display:grid) {
  .c0 {
    max-width: initial;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
      onClick={[Function]}
    >
      <img
        alt="Image Alt"
        className="c3"
        height={null}
        sizes={null}
        src=""
        srcSet={null}
        width={null}
      />
    </div>
  </div>
</div>
`;
